control-plane-operator/controllers/hostedcontrolplane/hostedcontrolplane_controller.go

```diff
@@ -140,150 +140,167 @@ func setConditionByType(conditions *[]hyperv1.HostedControlPlaneCondition, condi
 	err := r.Client.Get(ctx, req.NamespacedName, hostedControlPlane)
 	if err != nil {
 		if apierrors.IsNotFound(err) {
-			return ctrl.Result{}, nil
+			r.Log.Info("hostedcontrolplane not found, skipping reconcile", "name", req.NamespacedName)
+			return
 		}
-		return ctrl.Result{}, err
+		reconcileErr = fmt.Errorf("failed to get hostedcontrolplane: %w", err)
+		return
 	}
 
 	// Return early if deleted
 	if !hostedControlPlane.DeletionTimestamp.IsZero() {
 		if err := r.delete(ctx, hostedControlPlane); err != nil {
-			r.Log.Error(err, "failed to delete cluster")
-			return ctrl.Result{}, err
+			reconcileErr = fmt.Errorf("failed to delete hostedcontrolplane: %w", err)
+			return
 		}
 		if controllerutil.ContainsFinalizer(hostedControlPlane, finalizer) {
 			controllerutil.RemoveFinalizer(hostedControlPlane, finalizer)
 			if err := r.Update(ctx, hostedControlPlane); err != nil {
-				return ctrl.Result{}, fmt.Errorf("failed to remove finalizer from cluster: %w", err)
+				reconcileErr = fmt.Errorf("failed to remove finalizer from cluster: %w", err)
+				return
 			}
 		}
-		return ctrl.Result{}, nil
+		return
 	}
 
-	oldStatus := hostedControlPlane.Status.DeepCopy()
-
 	// Fetch the Cluster.
 	cluster, err := util.GetOwnerCluster(ctx, r.Client, hostedControlPlane.ObjectMeta)
 	if err != nil {
-		return ctrl.Result{}, err
+		reconcileErr = fmt.Errorf("failed to get hostedcontrolplane owner cluster: %w", err)
+		return
 	}
 
 	if cluster == nil {
 		r.Log.Info("Cluster Controller has not yet set OwnerRef")
-		return ctrl.Result{}, nil
+		return
 	}
 
 	if util.IsPaused(cluster, hostedControlPlane) {
 		r.Log.Info("HostedControlPlane or linked Cluster is marked as paused. Won't reconcile")
-		return ctrl.Result{}, nil
+		return
 	}
 
 	// Ensure the hostedControlPlane has a finalizer for cleanup
 	if !controllerutil.ContainsFinalizer(hostedControlPlane, finalizer) {
 		controllerutil.AddFinalizer(hostedControlPlane, finalizer)
 		if err := r.Update(ctx, hostedControlPlane); err != nil {
-			return ctrl.Result{}, fmt.Errorf("failed to add finalizer to hostedControlPlane: %w", err)
+			reconcileErr = fmt.Errorf("failed to add finalizer to hostedControlPlane: %w", err)
+			return
 		}
 	}
 
 	r.Log = r.Log.WithValues("cluster", cluster.Name)
 
-	var result ctrl.Result
 	// TODO (alberto):
 	// May be eventually just run a deployment with a CVO running a hostedControlPlane profile
 	// passing the hostedControlPlane.spec.version through?
 
 	if hostedControlPlane.Status.Ready {
 		r.Log.Info("Is ready")
-		return r.setAvailableCondition(ctx, hostedControlPlane, oldStatus, hyperv1.ConditionTrue, "AsExpected", "HostedControlPlane is ready", result, nil)
+		return
 	}
 
+	// Recalculate status. Use a defer to ensure we always try to update status
+	// if a change is detected. When a status update is tried, make sure to wrap
+	// any existing error.
+	updatedStatus := *hostedControlPlane.Status.DeepCopy()
```

Comment:
I'd rather avoid do this in a defer call but would call our func explicitly where needed for readability even though the code will be more verbose.

Also can we rely on `Patch` rather than `Update`?

E.g, the likes of:
We do at the top before mutating:

```
baseToPatch := client.MergeFrom(hostedControlPlane.Status.DeepCopy())
```

Then we have a func to mutate where relevant. Then in another one we persist where relevant:

```
if err := r.Client.Status().Patch(context.Background(), hostedControlPlane, baseToPatch); err != nil {
			return err
		}
```
This would only be needed to be called once if we want to encapsulate the reconcile function in its one method.
This is in essence similar to what we do here https://github.com/openshift/hypershift/blob/main/hypershift-operator/controllers/nodepool/nodepool_controller.go#L120-L139

Reply:
Maybe... I'm not actually that sold on either the nodepool example or the one in this PR. Basically we've turned reconcile into a contract-free blob of side-effects (both external and mutations to input) and then rely on a blanket patch to cover up the reality that our functions don't have clear inputs and outputs which can be reasoned about.

Going to come up with a few other examples for consideration...